### PR TITLE
Upgrade Gradle to use Groovy 3

### LINKED ISSUE #15566: Accept Groovy performance regressions
Gradle currently uses Groovy 2.5.x line which is no longer receiving feature updates.
Upgrade to Groovy 3 to benefit from the latest language features.

This is a prerequisite for Gradle to be able to run on JDK16.

Areas to investigate
- [x] LazyTest
- [x] GroovyDoc tests (due to classpath)
- [x] CodeNarcCompilationClasspathIntegrationTest
- [x] IncrementalGroovyProjectBuildIntegrationTest
- [x] BuildSrcSpockIntegrationTest
- [x] Init tests (GroovyLibraryInitIntegrationTest, ...)
- [x] GroovyCrossCompilationIntegrationTest
- [x] Look into problems using groovybuilder and init scripts
- [x] Groovy Parser issues (ModelDslRuleInputDetectionIntegrationSpec, StronglyTypedConfigurationAttributesResolveIntegrationTest/StringConfigurationAttributesResolveIntegrationTest (@big-guy looking into those)
-- https://github.com/apache/groovy/pull/1485 (to be in `3.0.8`)
- [x] Upgrade to a wrapper built on master
- [x] Look into performance regressions (`BuildSrcApiChangePerformanceTest` and `JavaFirstUsePerformanceTest`)

### PR DESCRIPTION
**Pull request by bmuskalla**
TODO

- [ ] Release notes
- [ ] ...

Fixes #15566

### CODE REVIEW

**Review by big-guy — APPROVED** [review]

File: subprojects/core/src/main/java/org/gradle/api/internal/DependencyClassPathProvider.java
Comment (bmuskalla) [reply subprojects/core/src/main/java/org/gradle/api/internal/DependencyClassPathProvider.java:91]: No, it's a new dependency of groovydoc (https://github.com/apache/groovy/blob/GROOVY_3_0_7/subprojects/groovy-groovydoc/build.gradle)

File: subprojects/language-groovy/src/integTest/groovy/org/gradle/groovy/GroovyDocStampsIntegrationTest.groovy
Comment (bmuskalla) [reply subprojects/language-groovy/src/integTest/groovy/org/gradle/groovy/GroovyDocStampsIntegrationTest.groovy:24]: Sorry, this wasn't intended to be on the PR, leftover from reproducing something. Reverted

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/mutator/ApplyNonAbiChangeToJavaSourceFileMutator.java
Comment (bmuskalla) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/mutator/ApplyNonAbiChangeToJavaSourceFileMutator.java:29]: That'd be my plan. I was holding off doing this right now to avoid being blocked on consuming a new profiler until we merge the groovy branch

File: subprojects/docs/src/snippets/antMigration/importBuild/groovy/build.gradle
Comment (bmuskalla) [reply subprojects/docs/src/snippets/antMigration/importBuild/groovy/build.gradle:26]: Done

File: subprojects/distributions-dependencies/build.gradle.kts
Comment (bmuskalla) [reply subprojects/distributions-dependencies/build.gradle.kts:152]: Thanks for catching

File: subprojects/docs/build.gradle
Comment (bmuskalla) [reply subprojects/docs/build.gradle:553]: No, leftovers from the initial wrapper build. Reverted

File: subprojects/model-core/src/main/java/org/gradle/model/internal/manage/binding/DefaultStructBindingsStore.java
Comment (bmuskalla) [reply subprojects/model-core/src/main/java/org/gradle/model/internal/manage/binding/DefaultStructBindingsStore.java:221]: Done.

File: subprojects/core/src/integTest/groovy/org/gradle/api/ApplyPluginIntegSpec.groovy
Comment (bmuskalla) [reply subprojects/core/src/integTest/groovy/org/gradle/api/ApplyPluginIntegSpec.groovy:182]: Oops, no. Fixed

**Comment by bmuskalla** [comment]
@bot-gradle test RfN

File: subprojects/enterprise/build.gradle.kts
Comment (ldaley) [inline subprojects/enterprise/build.gradle.kts:10]: This doesn't seem right. There are test fixtures that need this, but the main code shouldn't.

Can you share what happens without this?

Comment (bmuskalla) [reply subprojects/enterprise/build.gradle.kts:10]: Indeed, this was only used in the fixtures. Removing the dependency from here, thanks for catchign

**Pull request merged**
